### PR TITLE
Activities

### LINKED ISSUE #1165: event model
I propose an "events" model, belonging to a planting (and maybe other models)
examples include: Diseases noticed, treated. Pests culled. pruned. fertilzer etc.

This can be presented as a timeline for each planting.

### LINKED ISSUE #3451: Record soil conditioning per garden
Additives:
- Soil
- Compost
- Blood & Bone
- Biochar?
- Other unspecified organic matter
- Mulches (Pea Straw, Straw, Other)

Activities:
- Till
- Mulch
- Watering. (I can picture a home automation system with smart taps generating a lot of this data)

Would be useful to have a timeline over the years of activities I've done to my growing medium.

### PR DESCRIPTION
**Pull request by CloCkWeRX**
Fix https://github.com/Growstuff/growstuff/issues/1165
Fix https://github.com/Growstuff/growstuff/issues/3451 (generally)

Add a very basic "Activities" model, which can work on a Garden and/or Planting or just be a general todo.

image
image
image
image

Probably missing a few places in the UI where there should be links or counts, but aren't

- [x] No RSS feed yet
- [x] No iCal feed yet
- [ ] No tests yet

### CODE REVIEW

File: app/views/home/_harvests.html.haml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-%h2= t('.recently_harvested')
+%h2= t('.recent_activities')
```
Comment (CloCkWeRX) [inline app/views/home/_harvests.html.haml]:
```suggestion
%h2= t('.recently_harvested')
```

File: app/controllers/activities_controller.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      )
+    end
+
+    respond_with @activity
+  end
+
+  def edit
+    # the following are needed to display the form but aren't used
+    @gardens = @activity.owner.gardens.active.order_by_name
+  @plantings = @activity.owner.plantings.active
```
Comment (CloCkWeRX) [inline app/controllers/activities_controller.rb]:
```suggestion
    @plantings = @activity.owner.plantings.active
```

File: app/models/activity.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  include Ownable
+  include Finishable
+  include SearchActivities
+
+  belongs_to :garden, optional: true
+  belongs_to :planting, optional: true
+
+  friendly_id :activity_slug, use: %i(slugged finders)
+
+  CATEGORIES = ["General", "Weeding", "Soil Cultivation", "Fertilizing", "Pruning", "Watering"]
```
Comment (CloCkWeRX) [inline app/models/activity.rb]: Suggestions welcome for any extras here

File: app/views/activities/_description.haml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#{link_to event_model.category, event_model} - #{event_model.name}
+
+- if event_model.planting
+  %span on #{link_to event_model.planting,event_model.planting}
```
Comment (CloCkWeRX) [inline app/views/activities/_description.haml]:
```suggestion
  %span on #{link_to event_model.planting, event_model.planting}
```

File: app/views/activities/_description.haml [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+
+- if event_model.garden
+  %span in #{link_to event_model.garden,event_model.garden}
```
Comment (CloCkWeRX) [inline app/views/activities/_description.haml]:
```suggestion
  %span in #{link_to event_model.garden, event_model.garden}
```

File: app/views/plantings/_owner.haml [hunk 6]
```diff
@@ -3,7 +3,7 @@
     .row
       .col
         %h4
-          Planted by
+          Planned by
```
Comment (CloCkWeRX) [inline app/views/plantings/_owner.haml]:
```suggestion
          Planted by
```

File: app/views/plantings/index.html.haml [hunk 7]
```diff
@@ -1,4 +1,4 @@
-- content_for :title, title('plantings', @owner, @crop, @planting)
+- content_for :title, title('activities', @owner, @crop, @planting)
```
Comment (CloCkWeRX) [inline app/views/plantings/index.html.haml]:
```suggestion
- content_for :title, title('plantings', @owner, @crop, @planting)
```

**Comment by CloCkWeRX** [comment]
So many typos from VSCode lagging.

**Comment by CloCkWeRX** [comment]
```
rspec ./spec/views/gardens/show.html.haml_spec.rb:35 # gardens/show shows plantings on the garden page
rspec ./spec/views/gardens/show.html.haml_spec.rb:31 # gardens/show renders markdown in the description
rspec ./spec/views/gardens/show.html.haml_spec.rb:19 # gardens/show shows the location
rspec ./spec/views/gardens/show.html.haml_spec.rb:23 # gardens/show shows the area
rspec ./spec/views/gardens/show.html.haml_spec.rb:27 # gardens/show shows the description
rspec ./spec/requests/plantings_spec.rb:30 # Plantings with a member GET /members/x/plantings.ics works!
rspec ./spec/views/plantings/show.html.haml_spec.rb:53 # plantings/show shows photos
rspec ./spec/views/plantings/show.html.haml_spec.rb:84 # plantings/show no location set doesn't contain a () if no location is set
rspec ./spec/views/plantings/show.html.haml_spec.rb:80 # plantings/show no location set renders markdown in the description
rspec ./spec/views/plantings/show.html.haml_spec.rb:72 # plantings/show no location set renders the quantity planted
rspec ./spec/views/plantings/show.html.haml_spec.rb:76 # plantings/show no location set renders the description
rspec ./spec/views/plantings/show.html.haml_spec.rb:95 # plantings/show location set shows the member's location in parentheses
rspec ./spec/views/plantings/show.html.haml_spec.rb:66 # plantings/show shows a link to add photos 
rspec ./spec/views/plantings/show.html.haml_spec.rb:29 # plantings/show sunniness shows the sunniness 
rspec ./spec/views/plantings/show.html.haml_spec.rb:28 # plantings/show sunniness shows the sunniness 
rspec ./spec/views/plantings/show.html.haml_spec.rb:39 # plantings/show planted from shows planted_from 
rspec ./spec/views/plantings/show.html.haml_spec.rb:40 # plantings/show planted from shows planted_from 
rspec ./spec/views/plantings/show.html.haml_spec.rb:49 # plantings/show planted from shows planted_from if blank 
rspec ./spec/features/gardens/adding_gardens_spec.rb:7 # Gardens signed in is accessible
```

**Comment by CloCkWeRX** [comment]
```
  7) plantings/show shows photos
     Failure/Error: - if @current_activities.size.positive?
 ```
However, that should be an activerecord collection rather than nil all of the time.

File: app/views/gardens/show.html.haml [hunk 8]
```diff
@@ -56,10 +56,21 @@
...
             %p Nothing is currently planted here.
 
-    %section.companions
-      %h2 Suggestioned companions
-      - @suggested_companions.each do |companion|
-        = render 'crops/tiny', crop: companion
+    %section
+      %h2 Current activities in garden
+      .index-cards
+        - if @current_activities.size.positive?
```
Comment (CloCkWeRX) [inline app/views/gardens/show.html.haml]:
```suggestion
        - if @current_activities&.size&.positive?
```

File: app/views/plantings/show.html.haml [hunk 9]
```diff
@@ -64,6 +64,17 @@
         %a{name: 'seeds'}
         = render 'plantings/descendants', planting: @planting
 
+      %section
+        %h2 Current activities for planting
+        .index-cards
+          - if @current_activities.size.positive?
```
Comment (CloCkWeRX) [inline app/views/plantings/show.html.haml]:
```suggestion
          - if @current_activities&.size&.positive?
```

File: app/models/activity.rb (hunk 3 above)
Comment (cesy) [reply app/models/activity.rb]: Is it worth distinguishing types of fertilising? E.g. giving the tomatoes liquid food or giving them more compost are both technically fertilising. Would the latter be soil cultivation? The rest sounds good.

Comment (cesy) [reply app/models/activity.rb]: We've already tracked harvesting separately, but should harvesting be changed to fit in this activity model?

Comment (cesy) [reply app/models/activity.rb]: Brenda also mentioned in #1165 about treating diseases - what are the sensible activities for that? I think the categories we have here cover everything needed on #3451.

**Review by cesy — COMMENTED** [review]
Overall this seems like a good idea and on a brief skim read it looks sensible.

File: app/models/activity.rb (hunk 3 above)
Comment (CloCkWeRX) [reply app/models/activity.rb]: > harvesting be changed to fit in this activity model?

Maybe long term it should have an interface to this. Sort of does via timeline service.

Or these activities should have their own models whenever their attributes are unique. IE a watering event has one duration/volume/source/salinity/etc.
A harvest has one or more crops, volume

**Comment by CloCkWeRX** [comment]
(have a bit of a play on staging)

File: app/views/plantings/index.ics.erb [hunk 10]
```diff
@@ -37,6 +37,16 @@ cal.description = "Plantings by #{@owner.login_name}"
     todo.summary = "Harvest #{planting['crop_name']}"
 
     cal.add_todo(todo)
+
+    # To work around google calendar not supporting VTODO, mirror the above as a specific event
+    event = Icalendar::Todo.new
```
Comment (CloCkWeRX) [inline app/views/plantings/index.ics.erb]:
```suggestion
    event = Icalendar::Event.new
```

File: app/models/activity.rb (hunk 3 above)
Comment (CloCkWeRX) [reply app/models/activity.rb]: For diseases, let's add a generic 'treatment'.

That covers topical application of many, many things, as well as more general "I applied a growth aid kind of like fertilizer, but not technically one"

Comment (CloCkWeRX) [reply app/models/activity.rb]: Liquid vs compost vs solid tablet vs blood and bone, I would put "mix compost into soil" or "mulching" as general soil cultivation activities, with the others specifically fertilizing.

Comment (CloCkWeRX) [inline app/models/activity.rb]:
```suggestion
  CATEGORIES = ["General", "Weeding", "Soil Cultivation", "Fertilizing", "Pruning", "Topical Application/Treating", "Watering"]
```

File: spec/requests/plantings_spec.rb [hunk 11]
```diff
@@ -96,7 +96,7 @@
         expect(data[1]["Date added"]).to eq @predictable_planting.created_at.to_fs(:db)
         expect(data[1]["License"]).to eq "CC-BY-SA Growstuff http://growstuff.org/"
 
-        expect(data.count).to eq 6
+        expect(data.count).to eq 7
```
Comment (CloCkWeRX) [inline spec/requests/plantings_spec.rb]:
```suggestion
        expect(data.count).to eq 6
```

**Pull request merged**